From cb9cc2399ed7826bd463a54a47f43bf2e22509a8 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Thu, 9 Oct 2014 21:00:25 -0400
Subject: [fs] xfs: fix cil push sequence after log recovery

Message-id: <1412888441-4181-32-git-send-email-sandeen@redhat.com>
Patchwork-id: 97329
O-Subject: [RHEL7.1 PATCH 31/47] [fs] xfs: fix cil push sequence after log recovery
Bugzilla: 1145837
RH-Acked-by: Brian Foster <bfoster@redhat.com>
RH-Acked-by: Dave Chinner <dchinner@redhat.com>

Bugzilla: 1145837
Upstream Status: Committed upstream
Build Info: https://brewweb.devel.redhat.com/taskinfo?taskID=8082426
Tested: locally using xfstests

Thanks,
-Eric

    commit 5c18717ea25fcb87c746deac1bf2ef2ff46e154a
    Author: Mark Tinguely <tinguely@sgi.com>
    Date:   Thu Jul 24 20:49:40 2014 +1000

    xfs: fix cil push sequence after log recovery

    When the CIL checkpoint is fully written to the log, the LSN of the checkpoint
    commit record is written into the CIL context structure. This allows log force
    waiters to correctly detect when the checkpoint they are waiting on have been
    fully written into the log buffers.

    However, the initial context after mount is initialised with a non-zero commit
    LSN, so appears to waiters as though it is complete even though it may not have
    even been pushed, let alone written to the log buffers. Hence a log force
    immediately after a filesystem is mounted may not behave correctly, nor does
    commit record ordering if multiple CIL pushes interleave immediately after
    mount.

    To fix this, make sure the initial context commit LSN is not touched until the
    first checkpointis actually pushed.

    [dchinner: rewrite commit message]

    Signed-off-by: Mark Tinguely <tinguely@sgi.com>
    Reviewed-by: Dave Chinner <dchinner@redhat.com>
    Signed-off-by: Dave Chinner <david@fromorbit.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/xfs/xfs_log_cil.c b/fs/xfs/xfs_log_cil.c
index de835da..843a6f3 100644
--- a/fs/xfs/xfs_log_cil.c
+++ b/fs/xfs/xfs_log_cil.c
@@ -78,8 +78,6 @@ xlog_cil_init_post_recovery(
 {
  log->l_cilp->xc_ctx->ticket = xlog_cil_ticket_alloc(log);
  log->l_cilp->xc_ctx->sequence = 1;
- log->l_cilp->xc_ctx->commit_lsn = xlog_assign_lsn(log->l_curr_cycle,
-        log->l_curr_block);
 }
 
 /*
-- 
1.7.1